### PR TITLE
Make Client SDK work w/o Wasm

### PR DESCRIPTION
**Pull request by mitschabaude**
This PR adds a `js_backend`  which could replace the entire Wasm/Rust output with the minimal functionality needed to make the Client SDK and mina-signer work.

This could provide an alternative route to achieving smaller size and a simpler build in the client SDK, without uglifying the rest of the code base with "non-consensus" switches.

EDIT: Per @mrmr1993's recommendations, I updated this PR to not actually change the client sdk and mina-signer. It only:
* checks in the new js_backend, but doesn't yet use it
* refactors `random_oracle` so that it can be used with the pure ocaml implementation of poseidon, by introducing a virtual submodule `random_oracle.permutation`.
* applies very small changes to Pickles that remove its top-level _runtime_ dependency on kimchi_bindings. I left this in because I think it's just good practice to avoid top-level execution. Also, there's the hope of eventually freeing snarkyjs from top-level Wasm execution.

### CODE REVIEW

**Comment by mrmr1993** [comment]
Heads up: I've opened #10659 as a kind of 'part 1' for removing the optcomp without reimplementing the rust primitives in JS. I've listed off a strategy for getting the client SDK compiling on top of that [here](https://github.com/MinaProtocol/mina/pull/10659#issuecomment-1082606621).

**Comment by mitschabaude** [comment]
Another thing that might need review is that I changed the logic for deserializing a verification key. Instead of always recomputing all parts of it, including the optional `wrap_vk`, I only try to recompute `wrap_vk` and leave it `None` when that doesn't work. The reason is that this deserialization logic is also used in the client_sdk, where it's impossible and unnecessary to compute `wrap_vk`, which e.g. requires computing the SRS.

**Comment by mitschabaude** [comment]
updated the name `fake_backend` to `js_backend`, to reflect that this is not fake. It properly implements a subset of the crypto primitives in JS

File: src/lib/crypto/kimchi_bindings/js/js_backend/unimplemented.js
Comment (mitschabaude) [reply src/lib/crypto/kimchi_bindings/js/js_backend/unimplemented.js]: Thanks! I ended up eliminating any dependencies of the client sdk on the shifts, and deleting this file. I thought it's cleaner if stuff needed only for snarks aren't accidentally used in parts that don't need them. (I wasn't quite consistent with this.. I did implement the `domain_generator` since I already had everything available to compute it.)

Here's the commit that eliminates dependency on the shifts: https://github.com/MinaProtocol/mina/pull/10654/commits/22694fb406fdcf936cd65ec4308c5a6416413d38

**Review by bkase — APPROVED** [review]
Since we disabled the connection this should be good to land -- we really need something working asap to unblock a bunch of work, but we can just ship this enabled on a side branch that we don't merge in and call the version "unstable-do-not-use-in-production" or something for now and then can fix before incentivized testnet.

**Comment by bkase** [comment]
@mrmr1993 is this alright with you?

File: src/lib/crypto/kimchi_bindings/js/js_backend/finite_field.js
Comment (mrmr1993) [inline src/lib/crypto/kimchi_bindings/js/js_backend/finite_field.js:190]: This gives the incorrect value for `x[0] === 0n`.

File: src/lib/crypto/kimchi_bindings/js/js_backend/finite_field.js
Comment (mrmr1993) [inline src/lib/crypto/kimchi_bindings/js/js_backend/finite_field.js:195]: Likewise

File: src/lib/crypto/kimchi_bindings/js/js_backend/finite_field.js
Comment (mrmr1993) [inline src/lib/crypto/kimchi_bindings/js/js_backend/finite_field.js:221]: Combine this with the function below?

File: src/lib/crypto/kimchi_bindings/js/js_backend/finite_field.js
Comment (mrmr1993) [inline src/lib/crypto/kimchi_bindings/js/js_backend/finite_field.js:295]: Can we switch to a direct `===`?

File: src/lib/crypto/kimchi_bindings/js/js_backend/finite_field.js
Comment (mrmr1993) [inline src/lib/crypto/kimchi_bindings/js/js_backend/finite_field.js:300]: Likewise

File: src/lib/crypto/kimchi_bindings/js/js_backend/bigint_helpers.js
Comment (mitschabaude) [reply src/lib/crypto/kimchi_bindings/js/js_backend/bigint_helpers.js:21]: no dedicated test. it seems to be tested implicitly by a combination of things; if I change the encoding in both `to_bytes` and `of_bytes`, I always get errors like `Invalid_curve_point`, i.e. some hard coded byte sequence doesn't successfully deserialize. If I only change `to_bytes` but not `of_bytes`, then I get errors in a couple of client SDK / mina-signer tests, because signatures can't be verified (to_bytes is used in some to_base58_check functions)

File: src/lib/crypto/kimchi_bindings/js/js_backend/finite_field.js
Comment (mitschabaude) [reply src/lib/crypto/kimchi_bindings/js/js_backend/finite_field.js:295]: yeah. I guess the idea was to test equality modulo p.. but right now we keep the numbers within `[0, p)` anyway so shouldn't matter

File: src/lib/crypto/kimchi_bindings/js/js_backend/elliptic_curve.js
Comment (mitschabaude) [reply src/lib/crypto/kimchi_bindings/js/js_backend/elliptic_curve.js:9]: good point, I made an issue to address this at a later stage of the mina-signer -> JS port https://github.com/o1-labs/snarkyjs/issues/517
(don't want to waste time stitching a Rust-generated file into the JSOO build, I'll do it as soon as I can work with normal JS modules & imports)

**Pull request merged**
